grakn-engine/src/main/java/ai/grakn/engine/tasks/storage/LockingBackgroundTask.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Grakn - A Distributed Semantic Database
+ * Copyright (C) 2016  Grakn Labs Limited
+ *
+ * Grakn is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Grakn is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Grakn. If not, see <http://www.gnu.org/licenses/gpl.txt>.
+ */
+
+package ai.grakn.engine.tasks.storage;
+
+import ai.grakn.engine.lock.LockProvider;
+import ai.grakn.engine.tasks.BackgroundTask;
+import ai.grakn.engine.tasks.TaskCheckpoint;
+import mjson.Json;
+
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.Lock;
+import java.util.function.Consumer;
+
+/**
+ * <p>
+ *     Task that control whens locking tasks begin
+ * </p>
+ *
+ * <p>
+ *     This utility class is used to help with background tasks which need to acquire a lock before starting
+ * </p>
+ *
+ * @author alexandraorth, fppt
+ */
+public abstract class LockingBackgroundTask implements BackgroundTask{
+
+    @Override
+    public boolean start(Consumer<TaskCheckpoint> saveCheckpoint, Json configuration) {
+        Lock engineLock = LockProvider.getLock(getLockingKey());
+        try {
+            // Try to get lock for one second. If task cannot acquire lock, it should return successfully.
+            boolean hasLock = engineLock.tryLock(1, TimeUnit.SECONDS);
```

Comment:
To frame the below: If a task returns `true` from the start method, it is marked as completed. If it returns `false`, it is marked as stopped. `STOPPED` tasks are never re-submitted if they are recurring, but `COMPLETED` tasks are. 

Using `tryLock` here might be an issue. This task will wait for the lock for 1 second, if it cannot acquire it, it returns `true` and gets marked as `COMPLETED`. 

This is the use case that we have designed for in the `PostProcessingTask`, which is recurring. If the PP task cannot acquire the lock it returns `true`, gets marked as `COMPLETED` and is resubmitted. If it can acquire the lock, it runs the post processing, returns `false` and gets marked as `STOPPED`, never to run again. 

Im not sure if we designed for this same use case in the counting task, which I *think* is non-recurring. This means that most instances of counting task will never run if they cannot acquire the lock the first time.